akka-stream/src/main/scala/akka/stream/StreamRefs.scala

```diff
@@ -0,0 +1,92 @@
+trait SinkRef[In] {
+
+  /** Scala API: Get [[Sink]] underlying to this source ref. */
+  def sink(): Sink[In, NotUsed]
+  /** Java API: Get [[javadsl.Sink]] underlying to this source ref. */
+  final def getSink(): javadsl.Sink[In, NotUsed] = sink().asJava
+}
+
+/**
+ * See full documentation on [[SourceRef]].
+ */
+object SourceRef {
+  /** Implicitly converts a SourceRef to a Source. The same can be achieved by calling `.source` on the SourceRef itself. */
+  implicit def convertRefToSource[T](ref: SourceRef[T]): Source[T, NotUsed] =
+    ref.source
+}
+
+/**
+ * A SourceRef allows sharing a "reference" with others, with the main purpose of crossing a network boundary.
+ * Usually obtaining a SourceRef would be done via Actor messaging, in which one system asks a remote one,
+ * to share some data with it, and the remote one decides to do so in a back-pressured streaming fashion -- using a stream ref.
+ *
+ * To create a [[SourceRef]] you have to materialize the `Source` that you want to obtain a reference to by attaching it to a `Sink.sourceRef`.
+ *
+ * Stream refs can be seen as Reactive Streams over network boundaries.
+ * See also [[akka.stream.SinkRef]] which is the dual of a `SourceRef`.
+ *
+ * For additional configuration see `reference.conf` as well as [[akka.stream.StreamRefAttributes]].
+ */
+trait SourceRef[T] {
+  /** Scala API: Get [[Source]] underlying to this source ref. */
+  def source: Source[T, NotUsed]
+  /** Java API: Get [[javadsl.Source]] underlying to this source ref. */
+  final def getSource: javadsl.Source[T, NotUsed] = source.asJava
+}
+
+// --- exceptions ---
+
+final case class TargetRefNotInitializedYetException()
+  extends IllegalStateException("Internal remote target actor ref not yet resolved, yet attempted to send messages to it. This should not happen due to proper flow-control, please open a ticket on the issue tracker: https://github.com/akka/akka")
+
+final case class StreamRefSubscriptionTimeoutException(msg: String)
+  extends IllegalStateException(msg)
+
+final case class RemoteStreamRefActorTerminatedException(msg: String) extends RuntimeException(msg)
+final case class InvalidSequenceNumberException(expectedSeqNr: Long, gotSeqNr: Long, msg: String)
+  extends IllegalStateException(s"$msg (expected: $expectedSeqNr, got: $gotSeqNr)")
+
+/**
+ * Stream refs establish a connection between a local and remote actor, representing the origin and remote sides
+ * of a stream. Each such actor refers to the other side as its "partner". We make sure that no other actor than
+ * the initial partner can send demand/messages to the other side accidentally.
+ *
+ * This exception is thrown when a message is recived from a non-partner actor,
+ * which could mean a bug or some actively malicient behaviour from the other side.
+ *
+ * This is not meant as a security feature, but rather as plain sanity-check.
+ */
+final case class InvalidPartnerActorException(expectedRef: ActorRef, gotRef: ActorRef, msg: String)
```

Comment:
Is that what happens on double materialization? If yes, then it should point it out that stream refs are "single-shot".

Reply:
yeap